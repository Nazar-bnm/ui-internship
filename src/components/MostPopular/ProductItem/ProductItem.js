import React, { Component } from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { NavLink } from 'react-router-dom';

import Modal from '../../Modal';
import ProductImage from '../../ProductImage';
import ProductOrder from '../../ProductOrder';
import Heading from '../../Heading';
import {
  REMOVED_FROM_WISHLIST_NOTIFICATION,
  ADDED_TO_WISHLIST_NOTIFICATION
} from '../../../constants/notificationData';

import './ProductItem.scss';

const CN = 'product';

class ProductItem extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isHovered: false,
      isShowedModal: false,
      convertedPrice: 0
    };

    this.showModal = this.showModal.bind(this);
    this.removeModal = this.removeModal.bind(this);
    this.handleMouseEnter = this.handleMouseEnter.bind(this);
    this.handleMouseLeave = this.handleMouseLeave.bind(this);
    this.addToWishlistWithNotification = this.addToWishlistWithNotification.bind(this);
    this.removeFromWishlistNotification = this.removeFromWishlistNotification.bind(this);
    this.renderWishlistButton = this.renderWishlistButton.bind(this);
    this.renderQuickView = this.renderQuickView.bind(this);
    this.renderHoverView = this.renderHoverView.bind(this);
  }

  async componentDidMount() {
    this.getConvertedPrice();
  }

  componentDidUpdate(prevProps) {
    const { currencyKey } = this.props;
    if (currencyKey !== prevProps.currencyKey) {
      this.getConvertedPrice();
    }
  }

  getConvertedPrice() {
    const {
      product: {
        price
      }
    } = this.props;
    const { currencyKey } = this.props;
    const currencyPrice = (price * currencyKey).toFixed(0);
    this.setState({
      convertedPrice: currencyPrice
    });
  }

  showModal() {
    this.setState({
      isShowedModal: true
    });
  }

  removeModal() {
    this.setState({
      isHovered: false,
      isShowedModal: false
    });
  }

  handleMouseEnter() {
    this.setState({
      isHovered: true
    });
  }

  handleMouseLeave() {
    const { isShowedModal } = this.state;

    !isShowedModal
      && this.setState({
        isHovered: false
      });
  }

  addToWishlistWithNotification() {
    const {
      product: { _id },
      addToWishlist,
      showMessage
    } = this.props;

    addToWishlist(_id);
    showMessage(ADDED_TO_WISHLIST_NOTIFICATION);
  }

  removeFromWishlistNotification() {
    const {
      product: { _id },
      removeFromWishlist,
      showMessage
    } = this.props;

    removeFromWishlist(_id);
    showMessage(REMOVED_FROM_WISHLIST_NOTIFICATION);
  }

  renderWishlistButton() {
    const {
      wishlist,
      product: { _id }
    } = this.props;

    const isInWishlist = wishlist && wishlist.find((wishedItemId) => _id === wishedItemId);

    const removeOrAddToWishlist = () => (isInWishlist
      ? this.removeFromWishlistNotification(_id)
      : this.addToWishlistWithNotification(_id));

    return (
      <i
        className={cx('product__icon', 'icon', 'heart', 'large', {
          red: isInWishlist,
          outline: !isInWishlist
        })}
        onClick={removeOrAddToWishlist}
      />
    );
  }

  renderQuickView() {
    const {
      product: {
        images,
        _id
      }
    } = this.props;

    const imagesForQuickView = images.map(({ claudinaryId }) => ({
      src: `${process.env.IMAGE_URL}/${claudinaryId}`
    }));

    return (
      <Modal className={`${CN}-quick-view`} removeModal={this.removeModal}>
        <Heading title="Quick view" />
        <div className={`${CN}-quick-view-content`}>
          <ProductImage className={`${CN}-quick-view-content-images`} images={imagesForQuickView} verticalCarousel />
          <div className={`${CN}-quick-view-content-wrapper`}>
            <ProductOrder
              className={`${CN}-quick-view-content__product-order`}
              id={_id}
            />
            <NavLink
              to={`/product-details/${_id}`}
              className={`${CN}-quick-view-content__link`}
            >
              See more details
            </NavLink>
          </div>
        </div>
      </Modal>
    );
  }

  renderHoverView() {
    const {
      product: {
        _id, images, label, title, price, sizes
      }, symbol
    } = this.props;
    const { convertedPrice } = this.state;
    const sizesWithSpaces = sizes.join(', ');

    const { isHovered, isShowedModal } = this.state;
    const imageSrc = images.length && `${process.env.IMAGE_URL}/${images[0].claudinaryId}`;

    return (
      <>
        <NavLink to={`/product-details/${_id}`}>
          <div
            className={cx(`${CN}__img-wrapper`, {
              [`${CN}__img-wrapper--hovered`]: isHovered
            })}
          >
            <div className={cx(`${CN}__img-wrapper__flag`, label)} />
            <div className={`${CN}__img-wrapper__labels`}>{label}</div>

            <img
              alt="product"
              className={cx(`${CN}__img-wrapper__img`, {
                [`${CN}__img-wrapper__img--hovered`]: isHovered
              })}
              src={imageSrc}
            />
          </div>
        </NavLink>
        <div
          className={cx(`${CN}__title-wrapper`, {
            [`${CN}__title-wrapper__title`]: isHovered
          })}
        >
          <NavLink
            className={cx(`${CN}__title-wrapper-link`)}
            to={`/product-details/${_id}`}
          >
            <h2
              className={cx(`${CN}__title-wrapper-link__title`, {
                [`${CN}__title-wrapper-link__title--hovered`]: isHovered
              })}
            >
              {title}
            </h2>
          </NavLink>
          {!isHovered && (
            <h3 className={`${CN}__title-wrapper__price`}>
              <>
                {symbol}
              </>
              {`${convertedPrice}`}

            </h3>
          )}
          {isHovered && (
            <>
              <span className={`${CN}__title-wrapper`}>
                {`Sizes: ${sizesWithSpaces}`}
              </span>
              <div className={`${CN}__title-wrapper icons`}>
                {window.innerWidth > 720 && (
                  <i
                    className="eye icon product__icon"
                    onClick={this.showModal}
                  />
                )}
                {this.renderWishlistButton()}
              </div>
              {isShowedModal && this.renderQuickView()}
            </>
          )}
        </div>
      </>
    );
  }

  render() {
    return (
      <div
        className={`${CN} slide`}
        onMouseEnter={this.handleMouseEnter}
        onMouseLeave={this.handleMouseLeave}
      >
        {this.renderHoverView()}
      </div>
    );
  }
}

const ItemShape = PropTypes.shape({
  _id: PropTypes.string,
  images: PropTypes.array,
  label: PropTypes.string,
  title: PropTypes.string,
  price: PropTypes.number,
  sizes: PropTypes.array
});

ProductItem.propTypes = {
  product: ItemShape.isRequired,
  wishlist: PropTypes.array,
  addToWishlist: PropTypes.func.isRequired,
  removeFromWishlist: PropTypes.func.isRequired,
  showMessage: PropTypes.func.isRequired
};

ProductItem.defaultProps = {
  wishlist: []
};

export default ProductItem;
