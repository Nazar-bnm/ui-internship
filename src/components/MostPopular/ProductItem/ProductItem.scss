@import '../../../scss/global_styles';
@import '../../../scss/styles';

.product {
  display: inline-block;
  text-align: center;
  vertical-align: top;
  // width: 100%;
  max-width: 12rem;
  padding: 5px;
  white-space: normal;

  &__button {
    background: none;
    color: inherit;
    border: none;
    padding: 0;
    font: inherit;
    cursor: pointer;
    outline: inherit;
    text-align: center;
  }

  &__img-wrapper {
    position: relative;
    width: 100%;
    padding: 0px;
    -webkit-transition: padding 1s;

    &--hovered {
      width: 100%;
      transition: all 700ms ease-out;
      padding: 0px 20px 0 20px;
    -webkit-transition: padding 1s;
    }

    &__flag {
      width: 0;
      height: 0;
      border-style: solid;
      border-width: 60px 60px 0 0;
      border-color: transparent;
      position: absolute;
      z-index: 2;
    }

    &__labels {
      font-family: font(secondary);
      font-size: 0.8rem;
      position: absolute;
      z-index: 3;
      margin: 7px 5px;
      color: $white;
      text-transform: uppercase;
    }

    &__img {
      width: 100%;
    }
  }
  &__title-wrapper.icons{
    @include flex($direction: row, $justify: space-evenly);
    width: 100%;
  }

  &__title-wrapper {
    margin: 10px 0;
    width: 100%;

    .icons {
      margin-top: 10px;
      border-top: 3px solid $white;
      width: 100%;

      .product__icon {
        display: block;
        font-size: 1.3rem;
        @include flex($align: center, $justify: center);
        margin-top: 10px;
      }

      .product__icon:hover {
        color: $orange;
        transition: 0.3s;
      }

      // .icon {
      //   display: block;
      //   font-size: 1.3rem;
      //   @include flex($align: center, $justify: center);
      //   margin-top: 10px;
      // }

      // .icon:hover {
      //   color: $orange;
      //   transition: 0.3s;
      // }
    }

    &__title {
      font-family: font(secondary);
      font-size: 0.9rem;
      text-transform: uppercase;
      margin: 10px 0;

      &--hovered {
        color: $orange;
      }
    }

    &__price {
      font-family: font(primary);
      text-align: center;
      font-size: 0.7rem;
      font-weight: 300;
    }
  }

  .sale{
    border-color: $dark-red transparent transparent transparent;
  }

  .new {
    border-color: $black transparent transparent transparent;
  }

  &-quick-view {
    max-height: 60vh;
    max-width: 80vw;
    padding: 1rem 3.5rem 1.5rem;

    .heading {
      &__wrapper {
        margin: 0 !important;
        margin-bottom: 1rem;
        height: 2rem;
      }

      &__title {
        font-size: 1.5rem;
      }
    }

    &-content {
      display: flex;
      justify-content: space-evenly;
      height: 90%;

      .product-order {
        display: block;
        width: 100%;
      }

      .product-image {
        height: 100%;

        &__container {
          flex-wrap: nowrap;
          margin-top: 8px;
          max-width: 40%;
          padding-right: 1rem;
        }

        &__big-image-container {
          width: fit-content;
          max-height: 70%;
        }

        &__small-images-container {
          max-height: 30%;
        }

        &__small-image {
          height: 100%;
        }
      }

      &__wrapper {
        height: 100%;
      }

      .product-order {
        &__heading {
          font-size: 1.5rem;
          margin-top: 0;
        }

        &__dropdowns-wrapper {
          height: 7rem;
        }

        .dropdown {
          height: 2rem;
        }
      }

      &__link {
        display: flex;
        justify-content: flex-end;
        padding-top: 8px;
        margin-left: auto;
        color: $orange;

        &:active,
        &:hover {
          color: $dove-gray;
        }
      }

      .button {
        cursor: pointer;
        width: 120px;
        height: 40px;
        color: $white;
        background-color: $black;
        font-family: inherit;
        font-size: 0.8rem;
        text-transform: uppercase;

        &:hover {
          color: $black;
          background-color: $white;
        }
      }
    }
  }
}

@media screen and (max-width: 768px) {
  .quick-view {
    &-content {
      display: flex;
      height: 90%;

      .heading__title {
        font-size: 5rem;
      }
      
      .product-order__heading {
        font-size: 1rem;
        margin-top: 0;
      }
    }
  }

  .product {
    display: inline-block;
    padding: 50px;
  }
}

@include media-for(mobile-landscape) {
  .product {
    display: inline-block;
    padding: 30px;
  }
}