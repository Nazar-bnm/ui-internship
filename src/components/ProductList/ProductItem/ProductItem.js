// REMOVE THIS BEFORE COMMITING!!!

import React from 'react';
import PropTypes from 'prop-types';

const ProductItem = ({ item }) => {
  const { id, photo, title, collection, category, price } = item;
  const imageStyle = {
    width: '100px',
    height: '120px',
  };

  const containerStyle = {
    width: '150px',
    height: '250px',
    display: 'flex',
    flexDirection: 'column',
    margin: '10px',
    padding: '10px',
    backgroundColor: '#FFDFD3',
  };

  const textStyle = {
    fontSize: '12px'
  };

  return (
    <div className="product-item__container" style={containerStyle}>
      <h4>{title}</h4>
      <img src={photo}
        style={imageStyle }></img>
      <p style={textStyle}>id: {id}</p>
      <p style={textStyle}>collection: {collection}</p>
      <p style={textStyle}>category: {category}</p>
      <p style={textStyle}>price: {price}</p>
    </div>
  );
};

ProductItem.propTypes = {
  item: PropTypes.object,
};

ProductItem.defaultProps = {
  item: {},
};

export default ProductItem;
