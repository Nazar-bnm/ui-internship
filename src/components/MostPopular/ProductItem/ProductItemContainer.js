import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import { addToWishlist, removeFromWishlist } from '../../../actions/actionsWishlist';
import showMessage from '../../../actions/actionsNotifications';
import ProductItem from './ProductItem';

const mapStateToProps = ({ reducerWishlist, productsReducer }, state) => ({
  wishlist: reducerWishlist.wishlist,
  products: productsReducer.products,
  cartItems: state.cart
});

const mapDispatchToProps = (dispatch) => ({
  addToWishlist: bindActionCreators(addToWishlist, dispatch),
  removeFromWishlist: bindActionCreators(removeFromWishlist, dispatch),
  showMessage: bindActionCreators(showMessage, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(ProductItem);
