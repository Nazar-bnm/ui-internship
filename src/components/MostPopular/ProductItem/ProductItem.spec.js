import React from 'react';
import { shallow } from 'enzyme';

import ProductItem from './ProductItem';

describe('<ProductItem />', () => {
  let wrapper;
  let addToWishlist;
  let showMessage;
  const removeFromWishlist = jest.fn();

  const props = {
    product: {
      _id: '5',
      images: [],
      label: 'new',
      title: 'dress',
      price: '599',
      sizes: ['s', 'm', 'l']
    },
    wishlist: ['5', '4', '2'],
    addToWishlist: addToWishlist,
    removeFromWishlist: removeFromWishlist,
    showMessage: showMessage
  };

  beforeEach(() => {
    addToWishlist = jest.fn();
    showMessage = jest.fn();
    props.addToWishlist = addToWishlist;
    props.showMessage = showMessage;
    wrapper = shallow(<ProductItem {...props} />);
  });

  it('should render component <ProductItem />', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should onClick eye button set state to isShowedModal to true', () => {
    wrapper.setState({ isHovered: true });
    wrapper
      .find('.eye')
      .first()
      .simulate('click');

    expect(wrapper.state().isShowedModal).toBeTruthy();
  });

  it('should set isShowedModal to false when removeModal is called', () => {
    wrapper.setState({ isHovered: true, isShowedModal: true });
    wrapper
      .find('.product-quick-view')
      .first()
      .props()
      .removeModal();

    expect(wrapper.state().isShowedModal).toBeFalsy();
  });

  it('should onMouseEnter set isHovered to true', () => {
    wrapper
      .find('.product')
      .first()
      .simulate('mouseEnter');

    expect(wrapper.state().isHovered).toBeTruthy();
  });

  it('should onMouseLeave set isHovered to false when isShowedModal is false', () => {
    wrapper
      .find('.product')
      .first()
      .simulate('mouseLeave');
    wrapper.setState({ isShowedModal: true });

    expect(wrapper.state().isHovered).toBeFalsy();
  });

  it('should onClick on red heart call removeFromWishlist and showNotification props', () => {
    wrapper.setState({ isHovered: true });
    wrapper
      .find('.red')
      .first()
      .simulate('click');

    expect(removeFromWishlist).toHaveBeenCalledTimes(1);
    expect(showMessage).toHaveBeenCalledTimes(1);
  });

  it('should call props addToWishlist and showMessage', () => {
    const classNameSelector = '.product-quick-view-content__product-order';
    
    wrapper.setState({ isHovered: true, isShowedModal: true });
    wrapper
      .find(classNameSelector)
      .first()
      .props()
      .onClickAddToWishlist();

    expect(addToWishlist).toHaveBeenCalledTimes(1);
    expect(showMessage).toHaveBeenCalledTimes(1);
  });
});
