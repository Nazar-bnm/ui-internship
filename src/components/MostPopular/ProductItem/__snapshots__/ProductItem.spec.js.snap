// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> should render component <ProductItem /> 1`] = `
<div
  className="product slide"
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <NavLink
    to="/product-details/5"
  >
    <div
      className="product__img-wrapper"
    >
      <div
        className="product__img-wrapper__flag new"
      />
      <div
        className="product__img-wrapper__labels"
      >
        new
      </div>
      <img
        alt="product"
        className="product__img-wrapper__img"
        onError={[Function]}
        src="undefined/undefined"
      />
    </div>
  </NavLink>
  <div
    className="product__title-wrapper"
  >
    <NavLink
      className="product__title-wrapper-link"
      to="/product-details/5"
    >
      <h2
        className="product__title-wrapper-link__title"
      >
        dress
      </h2>
    </NavLink>
    <h3
      className="product__title-wrapper__price"
    >
      $599
    </h3>
  </div>
</div>
`;
